File: 2017/09_StreamProcessing/test/test_groups.py
# ======================================================================
# Stream Processing
#   Advent of Code 2017 Day 09 -- Eric Wastl -- https://adventofcode.com
#
# Computer simulation by Dr. Dean Earl Wright III
# ======================================================================

# ======================================================================
#                    t e s t _ g r o u p s . p y
# ======================================================================
"Test solver for Advent of Code 2017 day 9, Stream Processing"

# ----------------------------------------------------------------------
#                                                                 import
# ----------------------------------------------------------------------
import unittest

import groups

# ----------------------------------------------------------------------
#                                                              constants
# ----------------------------------------------------------------------

P1_GROUP_EXAMPLES = [
    "{}", # 1 group.
    "{{{}}}",  # 3 groups.
    "{{},{}}", # also 3 groups.
    "{{{},{},{{}}}}", # 6 groups.
    "{<{},{},{{}}>}", # 1 group (which itself contains garbage).
    "{<a>,<a>,<a>,<a>}", #1 group.
    "{{<a>},{<a>},{<a>},{<a>}}", # 5 groups.
    "{{<!>},{<!>},{<!>},{<a>}}", # 2 groups (since all but the last > are
]
P1_GROUP_NUMBERS = [1, 3, 3, 6, 1, 1, 5, 2]

P1_SCORE_EXAMPLES = [
    "{}", # score of 1.
    "{{{}}}", # score of 1 + 2 + 3 = 6.
    "{{},{}}", # score of 1 + 2 + 2 = 5.
    "{{{},{},{{}}}}", # score of 1 + 2 + 3 + 3 + 3 + 4 = 16.
    "{<a>,<a>,<a>,<a>}", # score of 1.
    "{{<ab>},{<ab>},{<ab>},{<ab>}}", # score of 1 + 2 + 2 + 2 + 2 = 9.
    "{{<!!>},{<!!>},{<!!>},{<!!>}}", # score of 1 + 2 + 2 + 2 + 2 = 9.
    "{{<a!>},{<a!>},{<a!>},{<ab>}}", # score of 1 + 2 = 3.
]
P1_SCORE_NUMBERS = [1, 6, 5, 16, 1, 9, 9, 3]

P2_EXAMPLES = [
    "{<>}", # 0 characters.
    "{<random characters>}", # 17 characters.
    "{<<<<>}", # 3 characters.
    "{<{!>}>}", # 2 characters.
    "{<!!>}", # 0 characters.
    "{<!!!>>}", # 0 characters.
    '{<{o"i!a,<{i<a>}', # 10 characters.
]
P2_NUMBERS = [0, 17, 3, 2, 0, 0, 10]

# ======================================================================
#                                                             TestGroups
# ======================================================================


class TestGroups(unittest.TestCase):  # pylint: disable=R0904
    """Test Register object"""

    def test_empty_init(self):
        """Test default Groups creation"""

        # 1. Create default Groups object
        mygrp = groups.Groups()

        # 2. Make sure it has the default values
        self.assertEqual(mygrp.part2, False)
        self.assertEqual(mygrp.state, 0)
        self.assertEqual(mygrp.level, 0)
        self.assertEqual(mygrp.group, 0)
        self.assertEqual(mygrp.score, 0)
        self.assertEqual(str(mygrp), "s=0 l=0 g=0 s=0")


    def test_text_init(self):
        """Test Groups creation from text"""

        # 1. Create Instruction object from text
        mygrp = groups.Groups(text="{<a>,<a>,<a>,<a>}")

        # 2. Make sure it has the specified values
        self.assertEqual(mygrp.part2, False)
        self.assertEqual(mygrp.state, 1)
        self.assertEqual(mygrp.level, 0)
        self.assertEqual(mygrp.group, 1)
        self.assertEqual(mygrp.score, 1)
        self.assertEqual(str(mygrp), "s=1 l=0 g=1 s=1")

    def test_part_one_groups(self):
        """Check the part one group examples"""

        # 1. Loop for all the part one group examples
        for enum, etext in enumerate(P1_GROUP_EXAMPLES):

            # 2. Create the group example
            mygrp = groups.Groups(text=etext)

            # 3. Verifty the results
            self.assertEqual(mygrp.group, P1_GROUP_NUMBERS[enum])

    def test_part_one_scores(self):
        """Check the part one score examples"""

        # 1. Loop for all the part one score examples
        for enum, etext in enumerate(P1_SCORE_EXAMPLES):

            # 2. Create the score example
            mygrp = groups.Groups(text=etext)

            # 3. Verifty the results
            self.assertEqual(mygrp.score, P1_SCORE_NUMBERS[enum])

    def test_part_two(self):
        """Check the part two examples"""

        # 1. Loop for all the part one score examples
        for enum, etext in enumerate(P2_EXAMPLES):

            # 2. Create the score example
            mygrp = groups.Groups(text=etext, part2=True)

            # 3. Verifty the results
            self.assertEqual(mygrp.trash, P2_NUMBERS[enum])

# ----------------------------------------------------------------------
#                                                  module initialization
# ----------------------------------------------------------------------
if __name__ == '__main__':
    pass

# ======================================================================
# end                    t e s t _ g r o u p s . p y                 end
# ======================================================================

File: 2017/09_StreamProcessing/test_groups.lua
-- ======================================================================
-- StreamProcessing
--   Advent of Code 2017 Day 09 -- Eric Wastl -- https://adventofcode.com
--
-- lua implementation by Dr. Dean Earl Wright III
-- ======================================================================

-- ======================================================================
--                    t e s t _ g r o u p s . l u a
-- ======================================================================
-- Test solver for Advent of Code 2017 day 09, StreamProcessing

-- ----------------------------------------------------------------------
--                                                                require
-- ----------------------------------------------------------------------
local luaunit = require("luaunit")

local groups = require("groups")

-- ----------------------------------------------------------------------
--                                                              from_text
-- ----------------------------------------------------------------------

function from_text(text)
  -- Break the text into trimed, non-comment lines

  -- 1. We start with no lines
  local result = {}

  -- 2. Loop for lines in the text
  for line in text:gmatch("[^\r\n]+") do

    -- 3. But ignore blank and comment lines
    line = line:gsub("%s*$", "")
    if #line > 0 and "!" ~= line:sub(1, 1) then

      -- 4. Add the line
      table.insert(result, line)
    end
  end

  -- 5. Return a table of cleaned lines
  return result
end

-- ----------------------------------------------------------------------
--                                                               dict_len
-- ----------------------------------------------------------------------
function dict_len(d)
    local result = 0
    for _ in pairs(d) do 
      result = result + 1
    end
    return result
end

-- ----------------------------------------------------------------------
--                                                              constants
-- ----------------------------------------------------------------------
local EXAMPLE_TEXT = [[
{{<a!>},{<a!>},{<a!>},{<ab>}}
]]
local PART_ONE_TEXT = EXAMPLE_TEXT
local PART_TWO_TEXT = EXAMPLE_TEXT

local PART_ONE_RESULT = 3
local PART_TWO_RESULT = 17

-- ======================================================================
--                                                             TestGroups
-- ======================================================================

function test_empty_init()
  -- Test the default Groups creation

  -- 1. Create default Groups object
  local myobj = groups:Groups()

  -- 2. Make sure it has the default values
  luaunit.assertEquals(myobj.part2, false)
  luaunit.assertEquals(#myobj.text, 0)

end

function test_text_init()
  -- Test the Groups object creation from text

  -- 1. Create Groups object from text
  local myobj = groups:Groups({text=from_text(EXAMPLE_TEXT)})

  -- 2. Make sure it has the expected values
  luaunit.assertEquals(myobj.part2, false)
  luaunit.assertEquals(#myobj.text, 1)

  -- 3. Check methods
  luaunit.assertEquals(myobj:next_state(1, "{"), 7)
  luaunit.assertEquals(myobj:next_state(1, "<"), 9)
  luaunit.assertEquals(myobj:next_state(1, ">"), 9)
  luaunit.assertEquals(myobj:next_state(1, "}"), 9)
  luaunit.assertEquals(myobj:next_state(1, ","), 9)
  luaunit.assertEquals(myobj:next_state(1, "!"), 9)
  luaunit.assertEquals(myobj:next_state(1, "a"), 9)
  luaunit.assertEquals(myobj:next_state(2, "{"), 7)
  luaunit.assertEquals(myobj:next_state(2, "<"), 3)
  luaunit.assertEquals(myobj:next_state(2, ">"), 9)
  luaunit.assertEquals(myobj:next_state(2, "}"), 8)
  luaunit.assertEquals(myobj:next_state(2, ","), 2)
  luaunit.assertEquals(myobj:next_state(2, "!"), 9)
  luaunit.assertEquals(myobj:next_state(2, "a"), 9)
  luaunit.assertEquals(myobj:next_state(3, "{"), 6)
  luaunit.assertEquals(myobj:next_state(3, "<"), 6)
  luaunit.assertEquals(myobj:next_state(3, ">"), 2)
  luaunit.assertEquals(myobj:next_state(3, "}"), 6)
  luaunit.assertEquals(myobj:next_state(3, ","), 6)
  luaunit.assertEquals(myobj:next_state(3, "!"), 4)
  luaunit.assertEquals(myobj:next_state(3, "a"), 6)
  luaunit.assertEquals(myobj:next_state(4, "{"), 3)
  luaunit.assertEquals(myobj:next_state(4, "<"), 3)
  luaunit.assertEquals(myobj:next_state(4, ">"), 3)
  luaunit.assertEquals(myobj:next_state(4, "}"), 3)
  luaunit.assertEquals(myobj:next_state(4, ","), 3)
  luaunit.assertEquals(myobj:next_state(4, "!"), 3)
  luaunit.assertEquals(myobj:next_state(4, "a"), 3)
  
  luaunit.assertEquals(myobj:score_braces("{}"), 1)
  luaunit.assertEquals(myobj:score_braces("{{{}}}"), 6)
  luaunit.assertEquals(myobj:score_braces("{{},{}}"), 5)
  luaunit.assertEquals(myobj:score_braces("{{{},{},{{}}}}"), 16)
  luaunit.assertEquals(myobj:score_braces("{<a>,<a>,<a>,<a>}"), 1)
  luaunit.assertEquals(myobj:score_braces("{{<ab>},{<ab>},{<ab>},{<ab>}}"), 9)
  luaunit.assertEquals(myobj:score_braces("{{<!!>},{<!!>},{<!!>},{<!!>}}"), 9)
  luaunit.assertEquals(myobj:score_braces("{{<a!>},{<a!>},{<a!>},{<ab>}}"), 3)
  
  luaunit.assertEquals(myobj:score_garbage("{}"), 0)
  luaunit.assertEquals(myobj:score_garbage("{{{}}}"), 0)
  luaunit.assertEquals(myobj:score_garbage("{{},{}}"), 0)
  luaunit.assertEquals(myobj:score_garbage("{{{},{},{{}}}}"), 0)
  luaunit.assertEquals(myobj:score_garbage("{<a>,<a>,<a>,<a>}"), 4)
  luaunit.assertEquals(myobj:score_garbage("{{<ab>},{<ab>},{<ab>},{<ab>}}"), 8)
  luaunit.assertEquals(myobj:score_garbage("{{<!!>},{<!!>},{<!!>},{<!!>}}"), 0)
  luaunit.assertEquals(myobj:score_garbage("{{<a!>},{<a!>},{<a!>},{<ab>}}"), 17)
  
  luaunit.assertEquals(myobj:score_garbage("{<>}"), 0)
  luaunit.assertEquals(myobj:score_garbage("{<random characters>}"), 17)
  luaunit.assertEquals(myobj:score_garbage("{<<<<>}"), 3)
  luaunit.assertEquals(myobj:score_garbage("{<{!>}>}"), 2)
  luaunit.assertEquals(myobj:score_garbage("{<!!>}"), 0)
  luaunit.assertEquals(myobj:score_garbage("{<!!!>>}"), 0)
  luaunit.assertEquals(myobj:score_garbage('{<{o"i!a,<{i<a>}'), 10)

end

function test_part_one()
  -- Test part one example of Groups object

  -- 1. Create Groups object from text
  local myobj = groups:Groups({text=from_text(PART_ONE_TEXT)})

  -- 2. Check the part one result
  luaunit.assertEquals(myobj:part_one({verbose=false, limit=0}), PART_ONE_RESULT)

end

function test_part_two()
  -- Test part two example of Groups object

  -- 1. Create Groups object from text
  local myobj = groups:Groups({part2=true, text=from_text(PART_TWO_TEXT)})

  -- 2. Check the part two result
  luaunit.assertEquals(myobj:part_two({verbose=false, limit=0}), PART_TWO_RESULT)

end

-- ----------------------------------------------------------------------
--                                                  module initialization
-- ----------------------------------------------------------------------
os.exit( luaunit.LuaUnit.run() )

-- ======================================================================
-- end                  t e s t _ g r o u p s . l u a                 end
-- ======================================================================
